pkg/tests/api/alerting/api_notification_channel_test.go

```diff
@@ -0,0 +1,463 @@
+	for _, alertName := range alertNames {
+		rules.Rules = append(rules.Rules, apimodels.PostableExtendedRuleNode{
+			GrafanaManagedAlert: &apimodels.PostableGrafanaRule{
+				Title:     alertName,
+				Condition: "A",
+				Data: []ngmodels.AlertQuery{
+					{
+						RefID: "A",
+						RelativeTimeRange: ngmodels.RelativeTimeRange{
+							From: ngmodels.Duration(time.Duration(5) * time.Hour),
+							To:   ngmodels.Duration(time.Duration(3) * time.Hour),
+						},
+						DatasourceUID: "-100",
+						Model: json.RawMessage(`{
+							"type": "math",
+							"expression": "2 + 3 > 1"
+						}`),
+					},
+				},
+			},
+		})
+	}
+
+	b, err := json.Marshal(rules)
+	require.NoError(t, err)
+
+	return string(b)
+}
+
+func TestNotificationChannels(t *testing.T) {
+	dir, path := testinfra.CreateGrafDir(t, testinfra.GrafanaOpts{
+		EnableFeatureToggles: []string{"ngalert"},
+		AnonymousUserRole:    models.ROLE_EDITOR,
+	})
+
+	store := testinfra.SetUpDatabase(t, dir)
+	grafanaListedAddr := testinfra.StartGrafana(t, dir, path, store)
+
+	mockChannel := newMockNotificationChannel(t, grafanaListedAddr)
+	amConfig := getAlertmanagerConfig(mockChannel.server.Addr)
+	rulesConfig := getRulesConfig(t)
+
+	channels.SlackAPIEndpoint = fmt.Sprintf("http://%s/slack_recvX/slack_testX", mockChannel.server.Addr)
+	channels.PagerdutyEventAPIURL = fmt.Sprintf("http://%s/pagerduty_recvX/pagerduty_testX", mockChannel.server.Addr)
+
+	// There are no notification channel config initially.
+	alertsURL := fmt.Sprintf("http://%s/api/alertmanager/grafana/config/api/v1/alerts", grafanaListedAddr)
+	_ = getRequest(t, alertsURL, http.StatusNotFound)
+
+	// Create the namespace we'll save our alerts to.
+	require.NoError(t, createFolder(t, store, 0, "default"))
+
+	// Post the alertmanager config.
+	u := fmt.Sprintf("http://%s/api/alertmanager/grafana/config/api/v1/alerts", grafanaListedAddr)
+	postRequest(t, u, amConfig, http.StatusAccepted)
+
+	// Verifying that all the receivers and routes have been registered.
+	resp := getRequest(t, alertsURL, http.StatusOK)
+	b, err := ioutil.ReadAll(resp.Body)
+	require.NoError(t, err)
+
+	cfg := apimodels.GettableUserConfig{}
+	require.NoError(t, json.Unmarshal(b, &cfg))
+
+	require.Equal(t, totalReceivers, len(cfg.AlertmanagerConfig.Receivers))
+	require.Equal(t, totalRoutes, len(cfg.AlertmanagerConfig.Route.Routes))
+
+	// Create rules that will fire as quickly as possible
+	u = fmt.Sprintf("http://%s/api/ruler/grafana/api/v1/rules/default", grafanaListedAddr)
+	postRequest(t, u, rulesConfig, http.StatusAccepted)
+
+	// Eventually, we'll get all the desired alerts.
+	// nolint:gosec
+	require.Eventually(t, func() bool {
+		fmt.Println(mockChannel.totalNotifications(), mockChannel.matchesExpNotifications(expNotifications))
+		return mockChannel.totalNotifications() == len(alertNames) &&
+			mockChannel.matchesExpNotifications(expNotifications)
+	}, 180*time.Second, 2*time.Second)
```

Comment:
3 minutes seems a bit too much 😓  -- Is this the soonest we can get it to consistently? I would expect this to be a max of 20s.

Reply:
It was taking 50-60s. I will see what settings needs to be tweaked to speed it up.

Reply:
It was the `"group_wait": "0s"` missing in the config for the routes. It take around 20s now, so setting this to 25s.